### PR TITLE
Unbeatable computer

### PR DESCRIPTION
**Pull request by whaley-ethan**

### CODE REVIEW

File: src/main/java/tic/tac/toe/Board.java [hunk 1]
```diff
@@ -30,6 +31,14 @@ public void setBoard(char[] board) {
     return emptyCells;
   }
 
+  public List<Character> getCells(int[] cellsToGet) {
+    List<Character> cells = Arrays.stream(cellsToGet)
+            .boxed()
+            .map(cell -> getMarkAt(cell))
```
Comment (ericmeyer) [inline src/main/java/tic/tac/toe/Board.java]: I'm not positive, but I think you can do something like this:
```suggestion
            .map(this::getMarkAt)
```

File: src/main/java/tic/tac/toe/Board.java [hunk 2]
```diff
@@ -30,6 +31,14 @@ public void setBoard(char[] board) {
     return emptyCells;
   }
 
+  public List<Character> getCells(int[] cellsToGet) {
+    List<Character> cells = Arrays.stream(cellsToGet)
```
Comment (ericmeyer) [inline src/main/java/tic/tac/toe/Board.java]:
```suggestion
    return Arrays.stream(cellsToGet)
```

File: src/main/java/tic/tac/toe/UnbeatableComputer.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            if (    board.isCellEmpty(threeCellsToCheck[cellIndex]) &&
+                    board.isCellEmpty(threeCellsToCheck[(cellIndex + 1) % 3]) &&
+                    board.getMarkAt(threeCellsToCheck[(cellIndex + 2) % 3]) == symbolToCheck) {
+                emptyCells = new int[] {threeCellsToCheck[cellIndex], threeCellsToCheck[(cellIndex + 1) % 3]};
+            }
+        }
+        return emptyCells;
+    }
+
+    int findWin(Board board, char symbolToCheck) {
```
Comment (whaley-ethan) [inline src/main/java/tic/tac/toe/UnbeatableComputer.java]: make explicit that functions are private

File: src/test/java/tic/tac/toe/MenuTest.java [hunk 4]
```diff
@@ -47,19 +47,14 @@ public void menuRepeatsAskingForPlayersWith3() {
     }
 
     @Test
-    public void menuTellsPlayerInvalidInputFor3Players() {
```
Comment (ericmeyer) [inline src/test/java/tic/tac/toe/MenuTest.java:50]: What caused this test to be removed?

**Review by ericmeyer — CHANGES_REQUESTED** [review]

File: src/main/java/tic/tac/toe/Game.java [hunk 5]
```diff
@@ -8,6 +8,7 @@
   private Player player2;
   private Player currentPlayer;
   private Player winningPlayer = null;
+  public static final int[][] WINNING_COMBINATIONS = new int[][]{{1, 4, 7}, {2, 5, 8}, {3, 6, 9}, {1, 2, 3}, {4, 5, 6}, {7, 8, 9}, {1, 5, 9}, {3, 5, 7}};
```
Comment (ericmeyer) [inline src/main/java/tic/tac/toe/Game.java]: If you recall, knowledge duplication between the `ComputerPlayer` and the `Game` was the original impetus to introduce the `Rules` class. Is there a way to leverage the `Rules` class you introduced to reduce the amount of complexity in `Game`?

File: src/test/java/tic/tac/toe/RulesTest.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
+package tic.tac.toe;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+/*
```
Comment (ericmeyer) [inline src/test/java/tic/tac/toe/RulesTest.java]: Can these comments be removed at this point?

File: src/test/java/tic/tac/toe/RulesTest.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+Other potential ideas:
+rules.hasWinner(board); // true / false
+rules.isOver(board); // true / false
+rules.isInProgress(board); // true / false
+rules.isTie(board); // true / false
+
+GameStatus.IN_PROGRESS / TIE_GAME / HAS_WINNER
+GameStatus.IN_PROGRESS / TIE_GAME / PLAYER_ONE_WINS
+ */
+public class RulesTest {
```
Comment (ericmeyer) [inline src/test/java/tic/tac/toe/RulesTest.java:7]: I would have expected more tests of this class. Ideally, this class would have been fully test-driven before updating the `ComputerPlayer` to use it.

For example, I would have expected tests for winning in a row, a column, a diagonal, etc.

**Review by ericmeyer — CHANGES_REQUESTED** [review]

**Review by ericmeyer — APPROVED** [review]

**Pull request merged**